pennylane/debugging/operation_checker.py

```diff
@@ -0,0 +1,1095 @@
+
+    The categorization of test results and of the associated messages is as follows:
+
+    - ``"pass"``: Status reports e.g., after a successfully completed run.
+
+    - ``"comment"``: Comments regarding certain properties of the tested operation(s).
+      These does not require any action to change the operation but is used to raise awareness
+      for behaviour that might be unexpected or differing from common operations.
+
+    - ``"hint"``: Remarks similar to warnings that indicate concrete hints to change the checked
+      operation(s). The recommended changes are expected to improve the code quality, performance
+      or consistency with other operations.
+
+    - ``"error"``: Problems with the checked operation(s) that require changes. These problems
+      might be in the core of the ``Operation``, preventing instantiation, or in a specific
+      method or property that is rendered unusable or inconsistent by the problem.
+
+
+    **Example**
+
+    Once instantiated, the ``OperationChecker`` can be called on an operation instance or class
+    with the following signature:
+
+    Args:
+
+      - op (type or .operation.Operation): Operation(s) to check, either provided as class
+        or as instance.
+
+      - parameters (Sequence[int or float]): Parameters with which the operation(s) is/are
+        expected to work. Ignored if ``op`` is not a type but a class instance. If not
+        provided, random parameters are used.
+
+      - wires (.wires.Wires): Wires with which the operation(s) is/are expected to work.
+        Ignored if ``op`` is not a type but a class instance. If not provided, consecutive
+        integer-labelled wires are used.
+
+      - seed (int): Seed for random generation of parameters.
+
+
+    Returns:
+
+      - str: The result status of the operation check, one of ``"error"``, ``"hint"``,
+        ``"comment"``, and ``"pass"``.
+      - str: A copy of the text printed via the ``print_fn`` during the check.
+
+    Let's consider a custom operation, which is a copy of ``qml.RX`` but with its
+    argument rounded to 2 decimal places:
+
+    .. code-block ::
+
+        class MyRX(qml.operation.Operation):
+
+            num_wires = 1
+
+            def __init__(self, theta, wires, do_queue=True, id=None):
+                theta = qml.math.round(theta, 2)
+                super().__init__(theta, wires=wires, do_queue=do_queue, id=id)
+
+            @staticmethod
+            def compute_matrix(theta):
+                theta = qml.math.round(theta, 2)
+                return qml.math.array([
+                    [qml.math.cos(theta / 2), -1j * qml.math.sin(theta / 2)],
+                    [-1j * qml.math.sin(theta / 2), qml.math.cos(theta / 2)],
+                ])
+
+    Now we can run the OperationChecker on this operation class:
+
+    >>> checker = qml.debugging.OperationChecker()
+    >>> result, output = checker(MyRX)
+    Checking operation MyRX for consistency.
```

Comment:
Cool idea!

I have one general worry: the checker has to closely co-evolve with the implementation details of the operator ecosystem. How do we make sure 

a) someone using it for tests, and it fails somehow due to the class logic not fitting the assumptions of the checker, doesn't need to now develop the checker class?
b) people stopping to use it after a short while, and it becomes a dead tool?
c) the checker not capturing something important that was added, and people having a false sense of security that it means all is good?